pkg/progressreader/progressreader.go

```diff
@@ -0,0 +1,52 @@
+package progressreader
+
+import (
+	"github.com/docker/docker/utils"
+	"io"
+)
+
+type StreamFormatter interface {
+	FormatProgress(string, string, *utils.JSONProgress) []byte
+	FormatStatus(string, string, ...interface{}) []byte
+	FormatError(error) []byte
+}
+
+// Reader with progress bar
+type Config struct {
+	In         io.ReadCloser // Stream to read from
+	Out        io.Writer     // Where to send progress bar to
+	Formatter  StreamFormatter
+	Size       int
+	LastUpdate int
+	NewLines   bool
+	ID         string
+	Action     string
+}
+
+func New(newReader Config) *Config {
+	return &newReader
+}
+func (config *Config) Read(p []byte) (n int, err error) {
+	read, err := config.In.Read(p)
+	config.Size += read
+	updateEvery := 1024 * 512 //512kB
+	if config.Size > 0 {
+		// Update progress for every 1% read if 1% < 512kB
+		if increment := int(0.01 * float64(config.Size)); increment < updateEvery {
+			updateEvery = increment
+		}
+	}
+	if config.Size-config.LastUpdate > updateEvery || err != nil {
+		config.Out.Write(config.Formatter.FormatProgress(config.ID, config.Action, &utils.JSONProgress{Total: config.Size, Current: config.Size}))
```

Comment:
I couldn't think of a non-hacky way to get around the fact that FormatProgress takes a *utils.JSONProgress. I got it to work by adding an int field to FormatProgress and passing a nil JSONProgress and config.Size in progressreader. The int field was useless for every other call to FormatProgress so I decided against committing that way of doing it.